onnxruntime/core/providers/qnn/builder/opbuilder/simple_op_builder.cc

```diff
@@ -153,14 +154,43 @@ Status SimpleOpBuilder::ProcessAlphaAttribute(QnnModelWrapper& qnn_model_wrapper
 
 // Support Transpose single node in QDQ model since it just change the data layout
 // Single node doesn't has any quantization parameters
-// Input tensors are created by previous node, output tensors created by next node
+// Input tensors are created by the previous node. Output tensors are created by the next node,
+// unless the output is the graph's final output.
 Status SimpleOpBuilder::HandleSingleTransposeNode(QnnModelWrapper& qnn_model_wrapper,
                                                   const NodeUnit& node_unit,
-                                                  std::vector<std::string>&& input_names) const {
+                                                  std::vector<std::string>&& input_names,
+                                                  bool is_quantized_model) const {
   std::vector<std::string> param_tensor_names;
   ORT_RETURN_IF_ERROR(ProcessPermAttribute(qnn_model_wrapper, node_unit, param_tensor_names));
+  const auto& outputs = node_unit.Outputs();
+  ORT_ENFORCE(outputs.size() == 1, "QNN Transpose node must have a single output.");
+  const auto& output = outputs[0];
+  auto& output_name = output.node_arg.Name();
+
+  const bool is_graph_output = qnn_model_wrapper.IsGraphOutput(output_name);
+
+  // Need to add output to the QNN model wrapper if this Transpose node's output is also
+  // the graph's output.
+  if (is_graph_output) {
+    const auto* type_proto = output.node_arg.TypeAsProto();
+    Qnn_DataType_t qnn_data_type = QNN_DATATYPE_UNDEFINED;
+    ORT_RETURN_IF_ERROR(GetQnnDataType(is_quantized_model, type_proto, qnn_data_type));
+
+    Qnn_QuantizeParams_t quantize_param = QNN_QUANTIZE_PARAMS_INIT;
+    std::vector<uint32_t> output_shape;
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(output.node_arg, output_shape),
+                                                     "Cannot get shape for QNN Transpose output");
+
+    Qnn_TensorType_t tensor_type = is_graph_output ? QNN_TENSOR_TYPE_APP_READ : QNN_TENSOR_TYPE_NATIVE;
+    QnnTensorWrapper output_tensorwrapper(output_name,
+                                          tensor_type,
+                                          qnn_data_type,
+                                          quantize_param,
+                                          std::move(output_shape));
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.AddTensorWrapper(std::move(output_tensorwrapper)),
+                                                         "Failed to add output tensor for QNN Transpose");
+  }
```

Comment:
The above fixes a bug when a Transpose node's output is also the graph's output.